### PR TITLE
Enables unloading of the cutter header from an attached trailer at th…

### PR DESCRIPTION
**Pull request by schwiti6190**
…e field.

Combine picks up the header of the trailer, which needs to be attached.
1) When the harvester is started from the hud, then the current position is used to store the trailer.
2) Starting from the AI menu uses the giants marker to store the trailer.

### CODE REVIEW

File: scripts/ai/AIDriveStrategyAttachHeader.lua [hunk 1]
```diff
@@ -0,0 +1,306 @@
+--[[
+This file is part of Courseplay (https://github.com/Courseplay/Courseplay_FS22)
+Copyright (C) 2022 Courseplay Dev Team
```
Comment (pvaiko) [inline scripts/ai/AIDriveStrategyAttachHeader.lua:3]: Year

File: scripts/ai/AIDriveStrategyAttachHeader.lua [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    else 
+        self.mode = self.MODES.ATTACH_HEADER_FROM_ATTACHED_TRAILER
+        self:debug("A cutter/header on a separate trailer is attached on the back.")
+        self.trailer = AIUtil.getImplementWithSpecialization(self.vehicle, DynamicMountAttacher)
+        self.dynamicMountAttacherController = DynamicMountAttacherController(vehicle, self.trailer)
+    end
+    self.attachableController = AttachableController(vehicle, self.trailer)
+    self.attacherJointController = AttacherJointController(vehicle, vehicle)
+    self:appendImplementController(self.attacherJointController)
+    local trailerAreaNode = createTransformGroup("tempTrailerNode")
```
Comment (pvaiko) [inline scripts/ai/AIDriveStrategyAttachHeader.lua:89]: Does not this needs to be removed after we don't need it?

File: scripts/ai/AIDriveStrategyAttachHeader.lua [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+    self.trailerAreaToAvoid =  PathfinderUtil.NodeArea(trailerAreaNode, -self.trailer.size.width/2, -self.trailer.size.length/2, self.trailer.size.width, self.trailer.size.length)
```
Comment (pvaiko) [inline scripts/ai/AIDriveStrategyAttachHeader.lua:90]: I might be wrong, and anyway, may not be a problem but this assumes that the cutter node is at the root node of the trailer and the root node is exactly in the middle of trailer (they sometimes have a lengthOffset for the size)

File: scripts/ai/AIDriveStrategyAttachHeader.lua [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+            goalNode = self.dynamicMountAttacherController:getCutterJointPositionNode()
+            rootNode = self.dynamicMountAttacherController:getMountedImplement().rootNode
+        else 
+            goalNode = self.attachableController:getCutterJointPositionNode()
+            rootNode = self.trailer.rootNode
+        end
+        local x, _, z = getWorldTranslation(goalNode)
+        local dirX, _, dirZ = localDirectionToWorld(rootNode, 0, 0, 1)
+        local rotY = MathUtil.getYRotationFromDirection(dirX, dirZ)
+        self.cutterNode = CpUtil.createNode("cutterNode", x, z, rotY)
```
Comment (pvaiko) [inline scripts/ai/AIDriveStrategyAttachHeader.lua:235]: Shouldn't this only create it if does not yet exist?

File: scripts/ai/controllers/AttachableController.lua [hunk 5]
```diff
@@ -0,0 +1,60 @@
+--- This controller is for implement, that can be attach to vehicles.
```
Comment (pvaiko) [inline scripts/ai/controllers/AttachableController.lua:1]: A spell check for these few lines maybe :)

File: scripts/ai/AIDriveStrategyAttachHeader.lua (hunk 3 above)
Comment (schwiti6190) [reply scripts/ai/AIDriveStrategyAttachHeader.lua:90]: A good point I am not even sure if this check is even needed. I guess I will put a comment there and we will see if it breaks from a user test.

Comment (pvaiko) [reply scripts/ai/AIDriveStrategyAttachHeader.lua:90]: Yeah, you right, the pathfinder should detect a collision there and avoid it anyway.

**Pull request merged**